src/libmongoc/doc/mongoc_structured_log_func_t.rst

```diff
@@ -0,0 +1,41 @@
+:man_page: mongoc_structured_log_func_t
+
+mongoc_structured_log_func_t
+============================
+
+Synopsis
+--------
+
+.. code-block:: c
+
+  typedef void (*mongoc_structured_log_func_t)
+  (const mongoc_structured_log_entry_t *entry, void *user_data);
+
+Callback function for :symbol:`mongoc_structured_log_set_handler`.
+Structured log handlers must be thread-safe.
+Logging may occur simultaneously on multiple application threads and/or background threads.
+
+Handlers may use any operating system or ``libbson`` functions but MUST not use any ``libmongoc`` functions except:
+
+* :symbol:`mongoc_structured_log_entry_get_component`
+* :symbol:`mongoc_structured_log_entry_get_level`
+* :symbol:`mongoc_structured_log_entry_message_as_bson`
+* :symbol:`mongoc_structured_log_get_level_name`
+* :symbol:`mongoc_structured_log_get_named_level`
+* :symbol:`mongoc_structured_log_get_component_name`
+* :symbol:`mongoc_structured_log_get_named_component`
+* :symbol:`mongoc_structured_log_get_max_level_for_component`
+
+Use of other ``libmongoc`` APIs will have undefined consequences, including deadlocks and unbounded recursion.
```

Comment:
We could detect this misuse, at a small performance penalty. It might be more practical once I try moving this from global to per-client, from Kevin's feedback. I'll see if we can add a guard that issues a warning at least.

Reply:
In the new non-global design this isn't nearly as big a problem- there's no immediate deadlock like the old design, and there are now ways to use libmongoc from inside a handler that are clearly correct, like using a different mongoc_client_t. I've changed the warning accordingly.